### PR TITLE
#1022 Command Migration: ('HEXISTS', 'HKEYS', 'HVALS')

### LINKED ISSUE #1022: Command Migration: ('HEXISTS', 'HKEYS', 'HVALS')
This issue tracks the migration of the mentioned commands - ('HEXISTS', 'HKEYS', 'HVALS') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by tarungka**
Migrates the eval functions for HEXISTS, HKEYS, HVALS to the new eval function type independent of protocol.
Closes #1022

Here is my check list:

- [x] Migrated the evalXXX function with the latest definition
- [x] Update or add unit tests for the new implementation.
- [x] All unit tests pass successfully.
- [x] Ensure all integration tests pass successfully.
- [x] Test that each protocol (RESP, HTTP, WebSocket) works correctly after migration.
- [x] Move Integration tests for the respective commands under the RESP integration tests directory from Async directory
- [x] Please validate that the documentation for the respective commands is up to date. If not then consider adding them.

### CODE REVIEW

File: internal/eval/store_eval.go [hunk 1]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+
+	key := args[0]
+	hmKey := args[1]
+	obj := store.Get(key)
+
+	var hashMap HashMap
+
+	if obj == nil {
+		return &EvalResponse{
+			Result: clientio.Encode(0, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:343]: Please send the response without encoding.

File: internal/eval/store_eval.go [hunk 2]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+			Result: nil,
+		}
+	}
+
+	hashMap = obj.Value.(HashMap)
+
+	_, ok := hashMap.Get(hmKey)
+	if ok {
+		return &EvalResponse{
+			Result: clientio.Encode(1, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:359]: Please send the response without encoding.

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  nil,
+		}
+	}
+	// Return 0, if specified field doesn't exist in the HashMap.
+	return &EvalResponse{
+		Result: clientio.Encode(0, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:365]: Please send the response without encoding.

File: internal/eval/store_eval.go [hunk 4]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+		if err := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeHashMap, object.ObjEncodingHashMap); err != nil {
+			return &EvalResponse{
+				Error:  diceerrors.ErrGeneral(diceerrors.WrongTypeErr),
+				Result: nil,
+			}
+		}
+		hashMap = obj.Value.(HashMap)
+	} else {
+		return &EvalResponse{
+			Result: clientio.Encode([]interface{}{}, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:401]: Please send the response without encoding.

File: internal/eval/store_eval.go [hunk 5]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+			Error:  nil,
+		}
+	}
+
+	for hmKey := range hashMap {
+		result = append(result, hmKey)
+	}
+
+	return &EvalResponse{
+		Result: clientio.Encode(result, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:411]: Please send the response without encoding. Raw results that makes it protocol independent.

File: internal/eval/store_eval.go [hunk 6]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+		return &EvalResponse{Error: diceerrors.ErrWrongArgumentCount("HVALS"), Result: nil}
+	}
+
+	key := args[0]
+	obj := store.Get(key)
+
+	if obj == nil {
+		// Return an empty array for non-existent keys
+		return &EvalResponse{
+			Result: clientio.Encode([]string{}, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:428]: Please send the response without encoding. Raw results that makes it protocol independent.

File: internal/eval/store_eval.go [hunk 7]
```diff
@@ -316,3 +316,136 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+
+	hashMap := obj.Value.(HashMap)
+	results := make([]string, 0, len(hashMap))
+
+	for _, value := range hashMap {
+		results = append(results, value)
+	}
+
+	return &EvalResponse{
+		Result: clientio.Encode(results, false),
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:448]: Please send the response without encoding. Raw results that makes it protocol independent.

File: test.log
Comment (AshwinKul28) [inline test.log:1]: Looks like this file got pushed by mistake. Please consider removing it.

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]
HI @tarungka thanks for the PR. please check the review comments and consider adding RESP integration tests as well.

**Comment by tarungka** [comment]
@AshwinKul28 I have updated the eval functions, will add the test cases shortly.

@soumya-codes I think you might have missed adding the item to the checklist, I'll check it off once added. Thanks.

**Comment by tarungka** [comment]
@AshwinKul28 I've added the integration test please review the PR, I will fix the conflicts if the changes look good.

File: internal/server/server.go [hunk 8]
```diff
@@ -267,6 +267,7 @@ func (s *AsyncServer) handleClientEvent(event iomultiplexer.Event) error {
 		return err
 	}
 
+	// function used within package, limit the scope
 	s.EvalAndRespond(commands, client)
```
Comment (tarungka) [inline internal/server/server.go:271]: @soumya-codes should this function be made private as it is only used within the package?

Comment (soumya-codes) [reply internal/server/server.go:271]: if it is used within the package, it should be private. However we will retire this code soon. So we may very well not put effort as well.

**Comment by AshwinKul28** [comment]
HI @tarungka I see that most of the integration tests are still under the async directory, please take an effort to move them under the resp directory. Thanks

**Comment by soumya-codes** [comment]
@tarungka please let us know if you have any concerns or facing any issues.

**Comment by tarungka** [comment]
@AshwinKul28 I have migrated the test cases, please review.

Apologies for the delay.

**Comment by apoorvyadav1111** [comment]
Hi @tarungka , Please resolve merge conflicts and let us know for a review. We are looking forward to close this soon.

Thanks

**Comment by apoorvyadav1111** [comment]
Hi @tarungka , Please review the unit test run logs. Thanks

**Comment by tarungka** [comment]
Hi @apoorvyadav1111, i have fixed almost all the test cases.

There is some new code recently merged, causing tests that return empty array responses to fail. I will look into that shortly.

**Comment by apoorvyadav1111** [comment]
Hi @tarungka, Sure, thanks. Please run unit, integration tests and linter after rebase with latest in local. If those pass, tag me and I will review and run workflow here.

File: integration_tests/commands/async/hkeys_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/async/hkeys_test.go:1]: if all the test cases for hkeys are migrated to resp, HTTP and websocket, we can delete this file

File: integration_tests/commands/async/hvals_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/async/hvals_test.go:1]: same as above

File: integration_tests/commands/http/hexists_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
+package http
+
+import (
+	"testing"
+	"time"
+
+	"gotest.tools/v3/assert"
```
Comment (apoorvyadav1111) [inline integration_tests/commands/http/hexists_test.go:7]: please migrate to github.com/stretchr/testify/assert package

File: integration_tests/commands/http/hexists_test.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+func TestHExists(t *testing.T) {
+	cmdExec := NewHTTPCommandExecutor()
+
```
Comment (apoorvyadav1111) [inline integration_tests/commands/http/hexists_test.go:12]: Please add a final cleanup

File: integration_tests/commands/http/hkeys_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/http/hkeys_test.go:1]: same as above

File: integration_tests/commands/http/hvals_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/http/hvals_test.go:1]: same as above

File: integration_tests/commands/resp/hexists_test.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+		{
+			name:     "RESP Check if field exists when no k,f and v exist",
+			commands: []string{"HEXISTS key field"},
+			expected: []interface{}{int64(0)},
+			debug:    true,
+		},
+		{
+			name:     "RESP HEXISTS operation against a key holding the wrong kind of value",
+			commands: []string{"SET key value", "HEXISTS key field"},
+			expected: []interface{}{"OK", "ERR -WRONGTYPE Operation against a key holding the wrong kind of value"},
```
Comment (apoorvyadav1111) [inline integration_tests/commands/resp/hexists_test.go:57]: Please confirm on this error message in redis for same test case

File: integration_tests/commands/resp/hexists_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/resp/hexists_test.go:1]: same as above

Comment (apoorvyadav1111) [inline integration_tests/commands/resp/hexists_test.go:1]: same as above

File: integration_tests/commands/resp/hkeys_test.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+		},
+		{
+			name:     "RESP HKEYS with non-existent key",
+			commands: []string{"HKEYS key_hkeys01"},
+			expected: []interface{}{[]interface{}{}},
+		},
+		{
+			name:     "RESP HKEYS with key containing a non-hash value",
+			commands: []string{"SET key_hkeys02 field1", "HKEYS key_hkeys02"},
+			expected: []interface{}{"OK", "ERR -WRONGTYPE Operation against a key holding the wrong kind of value"},
```
Comment (apoorvyadav1111) [inline integration_tests/commands/resp/hkeys_test.go:27]: please check response in redis for this scenario

File: integration_tests/commands/resp/hkeys_test.go
Comment (apoorvyadav1111) [inline integration_tests/commands/resp/hkeys_test.go:1]: same as other test files

File: integration_tests/commands/resp/hvals_test.go [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+		},
+		{
+			name:     "RESP HVALS with non-existing key",
+			commands: []string{"HVALS hvalsKey01"},
+			expected: []interface{}{[]any{}},
+		},
+		{
+			name:     "HVALS on wrong key type",
+			commands: []string{"SET hvalsKey02 field", "HVALS hvalsKey02"},
+			expected: []interface{}{"OK", "ERR -WRONGTYPE Operation against a key holding the wrong kind of value"},
```
Comment (apoorvyadav1111) [inline integration_tests/commands/resp/hvals_test.go:27]: same as above.

File: test.log
Comment (apoorvyadav1111) [inline test.log:1]: please remove this file

File: integration_tests/commands/resp/set_test.go [hunk 14]
```diff
@@ -12,6 +12,7 @@ type TestCase struct {
 	name     string
 	commands []string
 	expected []interface{}
+	debug    bool
```
Comment (apoorvyadav1111) [inline integration_tests/commands/resp/set_test.go:15]: why is this needed, please share

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]

**Comment by apoorvyadav1111** [comment]
Hi, I have left some comments, in addition to them, please use the "github.com/stretchr/testify/assert" in all test files related to the commands under the scope of this PR. Also all the tests need a final cleanup. Currently, we delete keys before test, therefore the last testcase have keys remaining which might affect the further tests.

File: integration_tests/commands/async/hkeys_test.go
Comment (tarungka) [reply integration_tests/commands/async/hkeys_test.go:1]: sure

File: integration_tests/commands/resp/hexists_test.go (hunk 11 above)
Comment (tarungka) [reply integration_tests/commands/resp/hexists_test.go:57]: image


This is the error message, the texts match.

**Comment by tarungka** [comment]
@AshwinKul28 @apoorvyadav1111 @lucifercr07 I have made the necessary changes, please review.

**Comment by AshwinKul28** [comment]
Hello @tarungka , hope you are doing well. All the codebase looks fantastic. I see there's still no documentation available for these commands. I know it's little back and forth but we want to close it once-for-all. Can you please consider [this template](https://github.com/DiceDB/dice/blob/master/docs/sample_command_docs.md) and try adding documentation pages for each of these three commands? Again, thanks a lot for the commendable efforts.

**Comment by tarungka** [comment]
Hi @AshwinKul28 sure, I will add these command docs. Do I create a markdown file for each of the three commands under docs/?

**Comment by AshwinKul28** [comment]
> Hi @AshwinKul28 sure, I will add these command docs. Do I create a markdown file for each of the three commands under docs/?

Yes please. thanks

**Comment by tarungka** [comment]
@AshwinKul28 @apoorvyadav1111 added the docs, please review.

**Review by apoorvyadav1111 — APPROVED** [review]

**Comment by apoorvyadav1111** [comment]
Amazing work! Thanks for finishing this up, will merge once the checks are successful.
